### PR TITLE
[3sat] extractor now included in zdf3sat.py (closes #21185)

### LINKED ISSUE #21185: [3sat] New URL
```
$ youtube-dl --ignore-config --verbose https://www.3sat.de/wissen/nano/nano-21-mai-2019-102.html
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--ignore-config', u'--verbose', u'https://www.3sat.de/wissen/nano/nano-21-mai-2019-102.html']                                                                                                                   
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.05.20
[debug] Python version 2.7.16 (CPython) - Linux-5.0.0-15-generic-x86_64-with-Ubuntu-19.04-disco
[debug] exe versions: ffmpeg 4.1.3, ffprobe 4.1.3, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[generic] nano-21-mai-2019-102: Requesting header
WARNING: Falling back on generic information extractor.
[generic] nano-21-mai-2019-102: Downloading webpage
[generic] nano-21-mai-2019-102: Extracting information
ERROR: Unsupported URL: https://www.3sat.de/wissen/nano/nano-21-mai-2019-102.html
Traceback (most recent call last):
  File "/home/marc/bin/youtube-dl/youtube_dl/extractor/generic.py", line 2340, in _real_extract
    doc = compat_etree_fromstring(webpage.encode('utf-8'))
  File "/home/marc/bin/youtube-dl/youtube_dl/compat.py", line 2551, in compat_etree_fromstring
    doc = _XML(text, parser=etree.XMLParser(target=_TreeBuilder(element_factory=_element_factory)))
  File "/home/marc/bin/youtube-dl/youtube_dl/compat.py", line 2540, in _XML
    parser.feed(text)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 1659, in feed
    self._raiseerror(v)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 1523, in _raiseerror
    raise err
ParseError: not well-formed (invalid token): line 134, column 841
Traceback (most recent call last):
  File "/home/marc/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 796, in extract_info
    ie_result = ie.extract(url)
  File "/home/marc/bin/youtube-dl/youtube_dl/extractor/common.py", line 529, in extract
    ie_result = self._real_extract(url)
  File "/home/marc/bin/youtube-dl/youtube_dl/extractor/generic.py", line 3329, in _real_extract
    raise UnsupportedError(url)
UnsupportedError: Unsupported URL: https://www.3sat.de/wissen/nano/nano-21-mai-2019-102.html

```

### PR DESCRIPTION
**Pull request by matthiasroos**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
With a [relaunch of website and medai library](http://www.quotenmeter.de/n/109532/eins-fuer-alles-3sat-legt-homepage-und-mediathek-zusammen) the extractor stopped working ([#21185](https://github.com/ytdl-org/youtube-dl/issues/21185)).
As correctly pointed out in the discussion of the issue by [barsnick](https://github.com/ytdl-org/youtube-dl/issues/21185#issuecomment-500046024), 3sat now relies on the infrastructure of its main contributor zdf. I added support for 3sat to the zdf extractor. A separate 3sat is no longer needed and therefore removed.

### CODE REVIEW

**Comment by matthiasroos** [comment]
I reverted my initial commit, because phoenix.py relies on dreisat.py, but is itself broken since a relaunch last year ([issue #17354](https://github.com/ytdl-org/youtube-dl/issues/17354)). Perhaps it is also possible to include phoenix in the zdf extractor (then as zdf3satphoenix.py), but it seems more complicated as for 3sat.

File: youtube_dl/extractor/extractors.py [hunk 1]
```diff
@@ -1522,6 +1522,6 @@
     ZattooIE,
     ZattooLiveIE,
 )
-from .zdf import ZDFIE, ZDFChannelIE
+from .zdf3sat import ZDFIE, ZDFChannelIE
```
Comment (dstftw) [inline youtube_dl/extractor/extractors.py]: Revert.

File: youtube_dl/extractor/zdf3sat.py [hunk 2]
```diff
@@ -39,7 +39,7 @@ def _extract_player(self, webpage, video_id, fatal=True):
 
 
 class ZDFIE(ZDFBaseIE):
-    _VALID_URL = r'https?://www\.zdf\.de/(?:[^/]+/)*(?P<id>[^/?]+)\.html'
+    _VALID_URL = r'https?://www\.(?:zdf|3sat)\.de/(?:[^/]+/)*(?P<id>[^/?]+)\.html'
```
Comment (dstftw) [inline youtube_dl/extractor/zdf3sat.py]: Not all URLs are interchangeable so it's most likely should be separate extractor.

File: youtube_dl/extractor/extractors.py (hunk 1 above)
Comment (matthiasroos) [reply youtube_dl/extractor/extractors.py]: Done.

File: youtube_dl/extractor/zdf3sat.py (hunk 2 above)
Comment (matthiasroos) [reply youtube_dl/extractor/zdf3sat.py]: Understood. Made new 3sat extractor based on zdf extractor.

**Comment by matthiasroos** [comment]
@dstftw Is it okay like that?

**Comment by dstftw** [comment]
No, it's not. Not copypasting here.

**Comment by matthiasroos** [comment]
But what do we gain by developing new code instead of using perfectly fine code, because it is not a coincidence, 3sat works with ZDF code.

1. 3sat and ZDF belong to the same network,
2. they both use the same website structure,
3. they both use the same zdf player,
4. they both rely on the same streaming service in the background,
5. there are precedences in your program for shared code (prosiebensat1 is one, in this case there are even more channels of one network crammed into one regex and so one extractor, each channel with a different URL) 

So in this light, what does still speak against the shared regex and extractor, except the different URLs? I'm just curious. If there are good arguments, I will develop a better solution without copy/paste.

**Comment by dstftw** [comment]
You **don't need** to develop a new code. You **must generalize and reuse** existing. Without copypasting.

**Comment by matthiasroos** [comment]
@dstftw How about now? I reuse the ZDF code in a separate extractor without any copy paste.

File: youtube_dl/extractor/zdf3sat.py (hunk 2 above)
Comment (FliegendeWurst) [reply youtube_dl/extractor/zdf3sat.py]: > Not all URLs are interchangeable so it's most likely should be separate extractor.

Interesting, would you mind giving an example?

File: youtube_dl/extractor/dreisat.py [hunk 3]
```diff
@@ -1,193 +1,29 @@
...
-    unified_strdate,
-    xpath_text,
-    determine_ext,
-    float_or_none,
-    ExtractorError,
-)
-
-
-class DreiSatIE(InfoExtractor):
-    IE_NAME = '3sat'
```
Comment (barsnick) [inline youtube_dl/extractor/dreisat.py]: The reworked extractor should probably keep the IE_NAME.

File: youtube_dl/extractor/dreisat.py [hunk 4]
```diff
@@ -1,193 +1,29 @@
+# coding: utf-8
 from __future__ import unicode_literals
 
-import re
+from .zdf import ZDFIE
+
+
+class DreiSatIE(ZDFIE):
+    _VALID_URL = r'https?://www\.3sat\.de/(?:[^/]+/)*(?P<id>[^/?]+)\.html'
+    _TESTS = [{
```
Comment (barsnick) [inline youtube_dl/extractor/dreisat.py]: If this passes the test, then: Nice! I have a different test case, but it's probably not worth the redundancy.

File: youtube_dl/extractor/dreisat.py [hunk 5]
```diff
@@ -1,193 +1,29 @@
...
+         'url': 'https://www.3sat.de/dokumentation/natur/dolomiten-sagenhaftes-juwel-der-alpen-100.html',
+         'info_dict': {
```
Comment (barsnick) [inline youtube_dl/extractor/dreisat.py]: If this passes the test, then: Nice! I have a different test case, but it's probably not worth the redundancy.

**Comment by basicmaster** [comment]
> But what do we gain by developing new code instead of using perfectly fine code, because it is not a coincidence, 3sat works with ZDF code.
> 
>     1. 3sat and ZDF belong to the same network,
> 
>     2. they both use the same website structure,
> 
>     3. they both use the same zdf player,
> 
>     4. they both rely on the same streaming service in the background,

Exactly. 3sat is a common program provided by ZDF (and ARD, ORF and SRG SSR). The technical processes like the linear TV channel, the website or these catch-up services are all maintained at ZDF in Mainz. So it would be absolutely reasonable to keep the name "ZDF" for the extractor; this would also keep the change smaller.

>     5. there are precedences in your program for shared code (prosiebensat1 is one, in this case there are even more channels of one network crammed into one regex and so one extractor, each channel with a different URL)

I agree - videos for the program SIXX (part of ProSiebenSAT.1) would also be downloaded by the prosiebensat1 extractor, though the name doesn't indicate any affiliation here.

**Pull request closed without merging**